### PR TITLE
send commands works directly with usernames

### LINKED ISSUE #9806: No need to request address for send command when recipient has ens username
# Bug Report

## Problem

When the recipient has an ens username, there is no need to ask for his address and the address pointed at by the ens name can be used instead.

### PR DESCRIPTION
**Pull request by yenda**
fix #9806 

Ready

### CODE REVIEW

**Review by vitvly — APPROVED** [review]

File: src/status_im/wallet/core.cljs [hunk 1]
```diff
@@ -375,21 +377,42 @@
...
+                                 :command? true
+                                 :value amount-hex}
+                                {:to       (ethereum/normalized-hex address)
+                                 :from     from-address
+                                 :chat-id  identity
+                                 :command? true
+                                 :data     (abi-spec/encode
+                                            "transfer(address,uint256)"
+                                            [to-norm amount-hex])
+                                 ;;Note: data from qr (eip681)
```
Comment (flexsurfer) [inline src/status_im/wallet/core.cljs]: there is no qr in chat , so this can be deleted

File: src/status_im/wallet/core.cljs [hunk 2]
```diff
@@ -498,19 +521,19 @@
...
-                                               :from     from-address
-                                               :data     (abi-spec/encode
-                                                          "transfer(address,uint256)"
-                                                          [to-norm amount-hex])
-                                               ;;Note: data from qr (eip681)
-                                               :gas      gas
-                                               :gasPrice gasPrice})})))))
+              (signing/sign
+               {:tx-obj (merge {:from  from-address
+                                ;;Note: data from qr (eip681)
```
Comment (flexsurfer) [inline src/status_im/wallet/core.cljs]: same here

File: src/status_im/wallet/core.cljs [hunk 3]
```diff
@@ -521,24 +544,45 @@
...
+                contact.db/enrich-contact))]
+    (cond-> {:db (assoc db
+                        :wallet/prepare-transaction
+                        {:from (ethereum/get-default-account
+                                (:multiaccount/accounts db))
+                         :to contact
+                         :symbol :ETH
+                         :from-chat? true})}
+      ens-verified
+      (assoc ::resolve-address
```
Comment (flexsurfer) [inline src/status_im/wallet/core.cljs:573]: it might take some time, and also there might be an error ,so we need to handle this

Comment (yenda) [reply src/status_im/wallet/core.cljs:573]: For now it is handled the same way as it is handled in the wallet

**Comment by churik** [comment]
Faced with very strange issue, however it is reproducible only in PR
### ISSUE 1: "Incoming transaction" message is not ftched from offline if it was sent after declining transaction and app was on background

*Steps:*
- create Account A , set ENS name A
- create Account B
- as Account B: create 1-1 with user A
- as Account B: request some ETH from user A
- as Account A: decline request
- as Account A: put app to background (for my phome always reproducible > 5 mins)
- as Account B: after 5 min again request some ETH from user A
- as Account A: go to foreground and wait for fetched messages

*Expected result:*
Account A received new ETH request
*Actual result:*
new ETH request does not arrive until relogin.
After relogin user receives new request 

Log before relogin (Account A): 
[no-request.log](https://github.com/status-im/status-react/files/4071094/no-request.log)

Log after relogin (Account A): 
[aft-relogin.log](https://github.com/status-im/status-react/files/4071095/aft-relogin.log)

The only difference is fleet - I'm testing your PR on eth.staging fleet, because it is not working with eth.prod

**Comment by churik** [comment]
Faced with very strange issue, however it is reproducible only in PR
### ISSUE 2: No "outgoing transaction" and "incoming transaction" messages if send assets to user with ENS name

*Steps:*
- create Account A , set ENS name A
- create Account B, set ENS name B
- as Account B: create 1-1 with user A
- as Account B: send some funds in 1-1 chat to user A

*Expected result:*
can see incoming and outgoing transaction messages in 1-1 chat
*Actual result:*
funds are sebt, but there are no messages in 1-1 chat about it

**Comment by yenda** [comment]
@churik it appears that your issues are related to being on a different fleet so I rebased. Also found a few bug related to failed transactions that are fixed now in the process.

**Comment by churik** [comment]
### ISSUE 3: When ENS name is still not set (but wallet has associated ENS) user will never receive assets

*Steps:*
- User A: restore multiaccount, that has assosiated ENS , but don't set ENS name
- User A: send message to public chat
- User B: restore any wallet with funds, join same public chat
- User B: tap on message from User A, Start 1-1 chat
- User B: try to send some ETH to User A 

*Expected result:*
full flow with requesting and sharing address
*Actual result:*
user can send transaction to non-existing address
So basically User B will lose his assets.
ident
User A  is not owner of address on screenshot.
non-exist

**Comment by churik** [comment]
### ISSUE 5: Unknown identifier of the user is shown in 'To' field if tap on "Sign and send" 

*Steps:*
- User A: restore multiaccount with funds
- User B: create new multiaccount, start chat with user A
- User B: request some ETH from user A
- User A: tap on Sign and Send

*Expected result:*
can see wallet address
*Actual result:*
some hash which is resolved to right address on tapping 'Send'
photo_2020-01-20 12 00 21 jpeg 2020-01-20 12-00-42

**Comment by churik** [comment]
### ISSUE 6: If "Send transaction" is sent after received messages, address is not requested and sent funds will be lost

Divide it in 2 scenarios, but consequences are the same.

#### Scenario 1
*Steps:*
- User B: create multiaccount
- User B: start chat with User A (also new user), send some messages
- User A:  try to send some assets  to User B


#### Scenario 2
Not possible to send ETH in 1-1 if chat after restoring multiaccount if had chatting with this user before 
*Steps:*
- User A: restore multiaccount,
- User B: create new account, send some messages to user A
- User A:  send some messages to User B
- uninstall app on both devices
- install app on both devices again and restore multiaccounts from seed phrase. On one of the devices chat will be recovered as well (in may be random, but in my case it was reproduced ~4 times)
- User A: if no chat with User B is recovered, start chat again
- User A: try to send some assets 


*Expected result:*
as User B will see message for sharing account
*Actual result:*
user can send transaction to a non-existing address
So basically User A  will lose his assets.

User B is not owner of address on screenshot.
photo_2020-01-20 13 43 25 jpeg 2020-01-20 13-43-49

**Comment by churik** [comment]
All related issues are fixed.
Tested flows:
- sent when ENS name is set, to main account
- request - sign and send when ENS name is set, to additional account
- flow for users without ENS
- flow for users on dofferent networks


Looks good to me.
